bottlecap/src/metrics/dogstatsd.rs

```diff
@@ -1,15 +1,36 @@
 use std::sync::mpsc::Sender;
 
 use crate::events::{self, Event, MetricEvent};
+use crate::metrics::aggregator::Aggregator;
+use crate::metrics::metric::Metric;
+use crate::metrics::datadog;
+use std::sync::{Arc, Mutex};
+use tracing;
 
 pub struct DogStatsD {
-    join_handle: std::thread::JoinHandle<()>,
+    serve_handle: std::thread::JoinHandle<()>,
+    aggregator: Arc<Mutex<Aggregator<1024>>>,
+    dd_api: datadog::DdApi,
 }
 
+pub struct DogStatsDConfig {
+    pub host: String,
+    pub port: u16,
+}
+const CONTEXTS: usize = 1024;
+
 impl DogStatsD {
-    pub fn run(host: &str, port: u16, event_bus: Sender<events::Event>) -> DogStatsD {
+    pub fn run(config: &DogStatsDConfig, event_bus: Sender<events::Event>) -> DogStatsD {
+        let aggr: Arc<Mutex<Aggregator<CONTEXTS>>> = Arc::new(Mutex::new(Aggregator::<CONTEXTS>::new().expect("failed to create aggregator")));
+        let serializer_aggr = Arc::clone(&aggr);
+        let serve_handle = DogStatsD::run_server(&config.host, config.port, event_bus, serializer_aggr);
+        let dd_api = datadog::DdApi::new();
+        DogStatsD { serve_handle, aggregator: aggr, dd_api }
+    }
+
+    fn run_server(host: &str, port: u16, event_bus: Sender<events::Event>, aggregator: Arc<Mutex<Aggregator<CONTEXTS>>>) -> std::thread::JoinHandle<()> {
         let addr = format!("{}:{}", host, port);
-        let join_handle = std::thread::spawn(move || {
+        std::thread::spawn(move || {
```

Comment:
FYI https://doc.rust-lang.org/std/thread/struct.Builder.html if you think that you might not be able to create a thread in your environment but need to handle that without panicing.

Reply:
We should be good here, but good to know!